src/_P021_Level.ino

```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
     case PLUGIN_TEN_PER_SECOND:
     {
-      // we're checking a var from another task, so calculate that basevar
-      taskIndex_t TaskIndex = P021_CHECK_TASK;
+      if (!P021_getConfigBit(P021_FLAGS, P021_SLOW_EVAL))
+      {
+        P021_evaluate(event);
+      }
+      success = true;
+      break;
+    }
+    case PLUGIN_ONCE_A_SECOND:
+    {
+      P021_check_autosafe(event); // This function relies on being called exactly once a second
 
-      if (!validTaskIndex(TaskIndex)) {
-        break;
+      if (P021_getConfigBit(P021_FLAGS, P021_SLOW_EVAL))
+      {
+        P021_evaluate(event);
       }
-      uint8_t BaseVarIndex = TaskIndex * VARS_PER_TASK + P021_CHECK_VALUE;
-      float   value        = UserVar[BaseVarIndex];
-      uint8_t state        = switchstate[event->TaskIndex];
-
-      // compare with threshold value
-      bool  isZero             = essentiallyZero(P021_TRIGGER_HYSTERESIS);
-      float valueLowThreshold  = P021_TRIGGER_LEVEL - (isZero ? 0.0f : (P021_TRIGGER_HYSTERESIS / 2.0f));
-      float valueHighThreshold = P021_TRIGGER_LEVEL + (isZero ? 1.0f : (P021_TRIGGER_HYSTERESIS / 2.0f)); // Include setvalue on
-                                                                                                          // 0-hysteresis
-
-      if (!definitelyGreaterThan(value, valueLowThreshold)) {
-        state = P021_INVERT_OUTPUT == 0 ? 1 : 0;
+      success = true;
+      break;
+    }
+  }
+
+  return success;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlState(int state)
+{
+  switch (state)
+  {
+    case P021_STATE_IDLE:     return F("Idle"); break;
+    case P021_STATE_HEATING:  return F("Heating"); break;
+    case P021_STATE_EXTEND:   return F("Extend"); break;
+    case P021_STATE_FORCE:    return F("Force"); break;
+    default:                  return F("***ERROR***"); break;
+  }
+}
+
+#endif // ifdef PLUGIN_021_DEBUG
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlMode(int mode)
+{
+  switch (mode)
+  {
+    case P021_OPMODE_CLASSIC: return F("Classic"); break;
+    case P021_OPMODE_OFF:     return F("Off"); break;
+    case P021_OPMODE_STANDBY: return F("Standby"); break;
+    case P021_OPMODE_ON:      return F("On"); break;
+    case P021_OPMODE_TEMP:    return F("Local"); break;
+    case P021_OPMODE_REMOTE:  return F("Remote"); break;
+    default:                  return F("***ERROR***"); break;
```

Comment:
No need for a `break` after `return`.

Reply:
Kind of defensive programming. Will remove it to conform to ESPeasy style.

Reply:
Once a `return` would fail to return we've got bigger issues to worry about 😆

Reply:
I learned (a very long time ago) to keep the complex coding structures as plain as possible as forgetting a break or other part of standard structures may lead to difficult to spot issues. Same teacher told to always add a default and never put returns halfway your code. The latter I do break :-) And for optimizations: "trust the compiler".  For embedded we often break all these rules. I adapted the code to ESPeasy style.

Reply:
Well, not to have _any_ code after a `return` is just a plain C++ (and many other languages) rule, not specific to ESPEasy 😉